src/main/java/frc/robot/scoreStates/ShootState.java

```diff
@@ -2,62 +2,139 @@
 
 import com.pathplanner.lib.auto.AutoBuilder;
 import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
 import edu.wpi.first.math.geometry.Translation2d;
 import edu.wpi.first.units.Units;
 import edu.wpi.first.wpilibj2.command.Command;
 import edu.wpi.first.wpilibj2.command.Commands;
 import frc.robot.Constants;
+import frc.robot.commandGroups.CommandGroups;
+import frc.robot.subsystems.hood.Hood;
+import frc.robot.subsystems.shooter.Shooter;
+import frc.robot.subsystems.shooter.ShooterConstants;
 import frc.robot.swerve.SwerveDrive;
 import java.util.List;
 import java.util.Set;
+import lib.PoseEstimation;
 import lib.Utils;
+import lib.math.interpolation.InterpolatingDouble;
 
 public class ShootState implements ScoreState {
+    private static Shooter shooter;
+    private static Hood hood;
+    private Translation2d speakerPose;
+    private InterpolatingDouble distanceToSpeaker;
+    private Pose2d botPose;
+    private List<Translation2d> optimalPoints;
+    private Translation2d optimalTranslation;
+    private double optimalRotation;
+    private static boolean inBounds;
+
+    private double shooterToSpeakerHeight =
+            ScoreStateConstants.SPEAKER_TARGET_HEIGHT
+                    - ShooterConstants.SHOOTER_HEIGHT.in(Units.Meters);
+
+    public ShootState() {
+        shooter = Shooter.getInstance();
+        hood = Hood.getInstance();
+    }
+
+    private Command rotate() {
+        return SwerveDrive.getInstance()
+                .turnCommand(
+                        optimalRotation, ScoreStateConstants.TURN_TOLERANCE.in(Units.Rotations));
+    }
 
     @Override
     public Command driveToClosestOptimalPoint() {
         return Commands.defer(
                 () -> {
-                    Translation2d speakerPose;
-                    Pose2d botPose = SwerveDrive.getInstance().getBotPose();
-                    List<Translation2d> optimalPoints;
+                    if (inBounds) {
+                        optimalTranslation = botPose.getTranslation();
+                        return rotate();
+                    }
+                    optimalTranslation = botPose.getTranslation().nearest(optimalPoints);
+
+                    return AutoBuilder.pathfindToPose(
+                                    new Pose2d(optimalTranslation, new Rotation2d(optimalRotation)),
+                                    Constants.AUTO_CONSTRAINTS)
+                            .andThen(rotate());
+                },
+                Set.of(SwerveDrive.getInstance()));
+    }
+
+    @Override
+    public Command initializeCommand() {
+        botPose = PoseEstimation.getInstance().getEstimatedPose();
+        return Commands.run(
```

Comment:
1. Use `runOnce()` instead of `run()`.
2. For the robot pose to update every run, line 68 needs to be moved inside the runnable.

Reply:
60d746efd05ec3ed4aacffed8fc4ceee8f57c93a